RFC-0027-PyTorch-Distributed-Tensor.md

```diff
@@ -0,0 +1,155 @@
+colwise_placement=[Shard(1)]  
+# distributed tensor returned will be sharded across the dimension specified in placements  
+distributed.empty((8, 12), device_mesh=device_mesh, placements=rowwise_placement)  
+  
+# if we want to do replication across a certain device list  
+replica_placement = [Replicate()]  
+# distributed tensor will be replicated to all four GPUs.  
+distributed.empty((8, 12), device_mesh=device_mesh, placements=replica_placement)  
+  
+# if we want to distributed a tensor with both replication and sharding  
+device_mesh = DeviceMesh(device_type="cuda", [[0, 1], [2, 3]])  
+# replicate across the first dimension of device mesh, then sharding on the second dimension of device mesh  
+spec=[Replicate(), Shard(0)]  
+distributed.empty((8, 8), device_mesh=device_mesh, placements=spec)  
+  
+# create a DistributedTensor that shards on dim 0, from a local torch.Tensor  
+local_tensor = torch.randn((8, 8), requires_grad=True)  
+rowwise_tensor = DTensor.from_local(local_tensor, device_mesh, rowwise_placement)  
+  
+# reshard the current rowise tensor to a colwise tensor or replicate tensor  
+colwise_tensor = rowwise_tensor.redistribute(device_mesh, colwise_placement)  
+replica_tensor = colwise_tensor.redistribute(device_mesh, replica_placement)
+
+```
+
+#### High level User Facing APIs
+
+Users can use DistributedTensor tensor constructors directly to create a distributed tensor (i.e. `distributed.ones/empty`), but for existing modules like nn.Linear that are already having torch.Tensor as parameters, how to make them distributed parameters? We offer a way to directly distribute a torch.Tensor and a module level APIs to directly distribute the module parameters. Below is the high level API we introduce:
+
+```python
+def distribute_tensor(tensor: torch.Tensor, device_mesh: DeviceMesh=None, placements: List[Placement]=None):  
+    '''
+    distribute the tensor according to device_mesh and placements, `tensor` could be a "meta" tensor.  
+    '''  
+  
+def distribute_module(  
+    module: nn.Module,  
+    device_mesh: DeviceMesh=None,  
+    partition_fn: Callable[str, nn.Module]=None,  
+    input_fn: Callable[torch.Tensor]=None,  
+    output_fn: Callable[torch.Tensor]=None,  
+):  
+    '''  
+    This function converts all module parameters to distributed tensor parameters according to the `partition_fn` specified.  
+    It could also control the input/output of the module by specifying the `input_fn` and `output_fn`. 
+    '''
+```
+
+#### High level API examples:
+
+```python 
+def MyModule(nn.Module):  
+    def __init__(self):  
+        super.__init__()  
+        self.fc1 = nn.Linear(8, 8)  
+        self.fc2 = nn.Linear(8, 8)  
+        self.relu = nn.ReLU()  
+     
+    def forward(self, input):  
+        return self.relu(self.fc1(input) + self.fc2(input))  
+  
+def shard_params(mod_name, mod):  
+    def to_dist_tensor(t): return distribute_tensor(t, mesh, rowwise_placement)  
```

Comment:
Do `mesh` and `rowwise_placement` need to be input to `shard_params`?

Reply:
`partition_fn` should have a fixed param signature list, I think `mesh` and `placements` are sth user can construct inside the `shard_params` function. I'll update it accordingly.